examples/onnx/cnn.py

```diff
@@ -0,0 +1,146 @@
+    '''
+    Converts a class vector (integers) to binary class matrix.
+    Args
+        y: class vector to be converted into a matrix
+            (integers from 0 to num_classes).
+        num_classes: total number of classes.
+    Return
+        A binary matrix representation of the input.
+    '''
+    y = np.array(y, dtype='int')
+    n = y.shape[0]
+    categorical = np.zeros((n, num_classes))
+    categorical[np.arange(n), y] = 1
+    categorical = categorical.astype(np.float32)
+    return categorical
+
+
+def preprocess(data):
+    data = data.astype(np.float32)
+    data /= 255
+    data = np.expand_dims(data, axis=1)
+    return data
+
+
+def accuracy(pred, target):
+    y = np.argmax(pred, axis=1)
+    t = np.argmax(target, axis=1)
+    a = y == t
+    return np.array(a, 'int').sum() / float(len(t))
+
+
+if __name__ == '__main__':
+
+
+    file_path = 'mnist.npz'
+    use_cpu = 'store_true'
+
+    print(file_path)
+    assert os.path.exists(file_path), \
+        'Pls download the MNIST dataset from https://s3.amazonaws.com/img-datasets/mnist.npz'
+
+    if use_cpu:
+        print('Using CPU')
+        dev = device.get_default_device()
+    else:
+        print('Using GPU')
+        dev = device.create_cuda_gpu()
+
+    train, test = load_data(file_path)
+    print(train[0].shape)
+    batch_number = 600
+    num_classes = 10
+    epochs = 1
+
+    sgd = opt.SGD(lr=0.01)
+
+    x_train = preprocess(train[0])
+    y_train = to_categorical(train[1], num_classes)
+
+    x_test = preprocess(test[0])
+    y_test = to_categorical(test[1], num_classes)
+    print('the shape of training data is', x_train.shape)
+    print('the shape of training label is', y_train.shape)
+    print('the shape of testing data is', x_test.shape)
+    print('the shape of testing label is', y_test.shape)
+
+    # operations initialization
+    conv1 = autograd.Conv2d(1, 32, 3, padding=1)
+    conv21 = autograd.Conv2d(32, 16, 3, padding=1)
+    conv22 = autograd.Conv2d(32, 16, 3, padding=1)
+    linear = autograd.Linear(32 * 28 * 28, 10)
+
+    def forward(x, t):
+        y = conv1(x)
+        y = autograd.tanh(y)
+        y1 = conv21(y)
+        y2 = conv22(y)
+        y = autograd.cat((y1, y2), 1)
+        y = autograd.sigmoid(y)
+        y = autograd.mul(y,y)
+        y = autograd.flatten(y)
+        y = linear(y)
+        loss = autograd.softmax_cross_entropy(y, t)
+        return loss, y
+
+    autograd.training = True
+    for epoch in range(epochs):
+        for i in range(batch_number):
+            inputs = tensor.Tensor(device=dev, data=x_train[
+                                   i * 100:(1 + i) * 100], stores_grad=False)
+            targets = tensor.Tensor(device=dev, data=y_train[
+                                    i * 100:(1 + i) * 100], requires_grad=False, stores_grad=False)
+
+            loss, y = forward(inputs, targets)
+
+            model = sonnx.get_onnx_model(loss, inputs, targets)
+            onnx.save(model, 'cnn.onnx')
```

Comment:
why do you save the model in every iteration?

Reply:
i took the function out from loop